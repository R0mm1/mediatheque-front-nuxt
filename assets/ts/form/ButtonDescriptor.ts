import AbstractDescriptor from '@/assets/ts/form/AbstractDescriptor'
import ButtonHrefDescriptor from '~/assets/ts/form/ButtonHrefDescriptor'

export type ButtonStyle = 'normal' | 'negative' | 'icon-round';
export type ButtonRoundedCorner = Boolean | Number; // Number to specify the value to use in CSS, true to use default value, false to deactivate

export default class ButtonDescriptor extends AbstractDescriptor {
  static descriptorType = 'ButtonDescriptor'

  static readonly typeButton = 'button'
  static readonly typeSubmit = 'submit'

  descriptorType = ButtonDescriptor.descriptorType
  type: ('button' | 'submit') = ButtonDescriptor.typeButton
  value: string = ''
  disabled: boolean = false
  customClasses: string[] = []
  style: ButtonStyle = 'normal'
  noDefaultStyle: boolean = false
  href?: ButtonHrefDescriptor = undefined

  constructor (name: string, value?: string, type?: string) {
    super(name)

    if (typeof value !== 'undefined') {
      this.value = value
    }

    if (typeof type !== 'undefined') {
      if (type !== ButtonDescriptor.typeButton && type !== ButtonDescriptor.typeSubmit) {
        throw new Error('Invalid type ' + type + ' provided')
      }
      this.type = type
    }
  }

  addCustomClass (customClass: string) {
    this.customClasses.push(customClass)
    return this
  }

  setValue (value: string) {
    this.value = value
    return this
  }

  setNoDefaultStyle (noDefaultStyle: boolean) {
    this.noDefaultStyle = noDefaultStyle
    return this
  }

  setHref (href: ButtonHrefDescriptor) {
    this.href = href
    return this
  }

  setStyle (style: ButtonStyle) {
    this.style = style
    return this
  }
}
