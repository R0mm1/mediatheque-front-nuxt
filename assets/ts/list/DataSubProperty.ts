/**
 * This class is meant to give indication about witch data to extract from the API's json response. It's to be used
 * with the Column class.
 */
export default class DataSubProperty {
    name: string;
    children: DataSubProperty[];

    // If we have multiple children, their values will be joined together with this string
    childrenJoinValueSeparator: string;

    // If a children give access to an array of object, for each object the values will extracted and joined with
    // childrenJoinValueSeparator and then each resulting string will be joined in one final string with this value
    childrenJoinCollectionSeparator: string;

    constructor (name: string, children: DataSubProperty[] = [], childrenJoinValueSeparator: string = ' ', childrenJoinCollectionSeparator: string = ', ') {
      this.name = name
      this.children = children
      this.childrenJoinValueSeparator = childrenJoinValueSeparator
      this.childrenJoinCollectionSeparator = childrenJoinCollectionSeparator
    }
}
