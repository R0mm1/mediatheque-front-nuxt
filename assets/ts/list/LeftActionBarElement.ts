import AbstractDescriptor from '@/assets/ts/form/AbstractDescriptor'

export default class LeftActionBarElement {
    callback!: Function;
    elementDescriptor!: AbstractDescriptor;

    constructor (callback: Function, elementDescriptor: AbstractDescriptor) {
      this.callback = callback
      this.elementDescriptor = elementDescriptor
    }
}
