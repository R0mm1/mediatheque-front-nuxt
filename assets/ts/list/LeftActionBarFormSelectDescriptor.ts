import SelectDescriptor from '@/assets/ts/form/SelectDescriptor'

export default class LeftActionBarFormSelectDescriptor extends SelectDescriptor {
  static descriptorType = 'LeftActionBarFormSelectDescriptor';

  descriptorType = LeftActionBarFormSelectDescriptor.descriptorType;

  constructor (name: string, options: { [index: string]: string }, label: string = '') {
    super(name, label)

    this.options = options
    this.noLabel = true

    return this
  }
}
