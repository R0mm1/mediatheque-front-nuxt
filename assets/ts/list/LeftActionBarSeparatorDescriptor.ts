import AbstractDescriptor from '~/assets/ts/form/AbstractDescriptor'

export default class LeftActionBarSeparatorDescriptor extends AbstractDescriptor {
  descriptorType = 'LeftActionBarSeparatorDescriptor';
}
