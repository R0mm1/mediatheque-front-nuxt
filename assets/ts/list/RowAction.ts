export default class RowAction {
  id: string
  label: string
  iconClassname: string
  confirmMessage: string = ''
  needConfirm: boolean = false

  /**
     * Define if the action button has to be displayed. The parameter gave to the method is the json object
     * corresponding to the row data
     * @param object
     */
  isDisplayed: Function = (object: object) => true

  constructor (id: string, label: string, iconClassname: string) {
    this.id = id
    this.label = label
    this.iconClassname = iconClassname
  }

  setNeedConfirm (needConfirm: boolean, confirmMessage: string = '') {
    this.needConfirm = needConfirm

    if (this.needConfirm) {
      if (confirmMessage.length === 0) { throw 'No confirmation message provided' }
      this.confirmMessage = confirmMessage
    }

    return this
  }

  setIsDisplayed (isDisplayed: Function) {
    this.isDisplayed = isDisplayed
    return this
  }
}
