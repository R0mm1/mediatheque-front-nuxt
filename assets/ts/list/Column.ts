import DataSubProperty from '@/assets/ts/list/DataSubProperty'

export type ColumnSortASC = 'ASC'
export type ColumnSortDESC = 'DESC'
export type ColumnSortNone = 'none'
export type ColumnSort = (ColumnSortASC | ColumnSortDESC | ColumnSortNone)

/**
 * Represent a column of the list
 */
export default class Column {
  static readonly sortUp: ColumnSortASC = 'ASC'
  static readonly sortDown: ColumnSortDESC = 'DESC'
  static readonly sortNone: ColumnSortNone = 'none'

  uid: string
  dataField: string
  label: string
  isSortable: boolean
  isSearchable: boolean
  subProperties: DataSubProperty[] = []
  searchParameterName: string
  orderParameterName: string
  sortState: ColumnSort = Column.sortNone
  searchString: string = ''

  /**
   * Same behavior than {@link DataSubProperty.childrenJoinValueSeparator}
   */
  joinValueSeparator: string = ' '

  /**
   * Same behavior than {@link DataSubProperty.childrenJoinCollectionSeparator}
   */
  joinCollectionSeparator: string = ', '

  constructor (dataField: string, label: string, isSortable: boolean = true, isSearchable: boolean = true) {
    // The dataField may be identical between two columns if subproperties are used, so we need to let the possibility
    // to override the way the column is identified, but in most cases the dataField should be enough
    this.uid = dataField
    this.dataField = dataField
    this.label = label
    this.isSortable = isSortable
    this.isSearchable = isSearchable

    // By default, we consider the API use the field name for the search and the order
    this.searchParameterName = dataField
    this.orderParameterName = dataField
  }

  setSubProperties (subProperties: DataSubProperty[]) {
    this.subProperties = subProperties
    return this
  }

  setSearchParameterName (searchParameterName: string) {
    this.searchParameterName = searchParameterName
    return this
  }

  setOrderParameterName (orderParameterName: string) {
    this.orderParameterName = orderParameterName
    return this
  }

  setSortStateFromString (sortState: string) {
    if (sortState !== Column.sortNone && sortState !== Column.sortDown && sortState !== Column.sortUp) {
      throw new Error('Invalid sort state ' + sortState)
    }
    this.sortState = sortState
  }

  setUid (uid: string) {
    this.uid = uid
    return this
  }

  setIsSortable (isSortable: boolean) {
    this.isSortable = isSortable
    return this
  }
}
