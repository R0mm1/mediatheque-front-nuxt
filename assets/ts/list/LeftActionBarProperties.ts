import LeftActionBarElement from '@/assets/ts/list/LeftActionBarElement'
import LeftActionBarSeparator from '@/assets/ts/list/LeftActionBarSeparator'

export default class LeftActionBarProperties {
    hasAddButton: boolean;

    customElements: (LeftActionBarElement | LeftActionBarSeparator)[];

    constructor (customElements: (LeftActionBarElement | LeftActionBarSeparator)[] = [], hasAddButton: boolean = true) {
      this.customElements = customElements
      this.hasAddButton = hasAddButton
    }
}
