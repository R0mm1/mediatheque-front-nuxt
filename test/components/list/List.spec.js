import 'reflect-metadata'
import VueRouter from 'vue-router'
import List from '@/components/list/List'
import { createLocalVue, mount } from '@vue/test-utils'
import '~/plugins/vueFormulate'
import { container } from 'tsyringe'
import Column from '~/assets/ts/list/Column'
import DataSubProperty from '~/assets/ts/list/DataSubProperty'
import LeftActionBarProperties from '~/assets/ts/list/LeftActionBarProperties'
import LeftActionBarElement from '~/assets/ts/list/LeftActionBarElement'
import LeftActionBarSeparatorDescriptor from '~/assets/ts/list/LeftActionBarSeparatorDescriptor'
import ButtonDescriptor from '~/assets/ts/form/ButtonDescriptor'
import LeftActionBarFormSelectDescriptor from '~/assets/ts/list/LeftActionBarFormSelectDescriptor'
import RowAction from '~/assets/ts/list/RowAction'
import BookService from '~/assets/ts/service/BookService'
import AuthenticationService from '~/assets/ts/service/AuthenticationService'
import RequestService from '~/assets/ts/service/RequestService'

jest.mock('~/assets/ts/objects/Navigator')
jest.mock('~/assets/ts/service/AuthenticationService', () => {
  return jest.fn().mockImplementation(() => {
    return {
      isLoggedIn: jest.fn().mockReturnValue(true),
      getToken: jest.fn().mockReturnValue('123')
    }
  })
})

const axios = require('axios')
const MockAdapter = require('axios-mock-adapter')
const axiosMock = new MockAdapter(axios)

describe('List', () => {
  const authenticationService = container.resolve(AuthenticationService)
  container.registerInstance(AuthenticationService, authenticationService)

  axiosMock.onGet(/.*\/books$/).reply(200, {
    '@context': '/api/contexts/Book',
    '@id': '/api/books',
    '@type': 'hydra:Collection',
    'hydra:member': [
      {
        '@id': '/api/electronic_books/39',
        '@type': 'ElectronicBook',
        id: 39,
        title: 'Sharko',
        year: '2017',
        language: 'Français',
        authors: [
          {
            '@id': '/api/authors/1',
            '@type': 'Author',
            id: 1,
            firstname: 'Franck',
            lastname: 'Thilliez'
          }
        ],
        shortSummary: 'Lorem ipsum'
      },
      {
        '@id': '/api/paper_books/45',
        '@type': 'PaperBook',
        id: 45,
        title: 'Sous les vents de Neptune',
        year: '2004',
        language: 'Français',
        authors: [
          {
            '@id': '/api/authors/5',
            '@type': 'Author',
            id: 5,
            firstname: 'Fred',
            lastname: 'Vargas'
          }
        ],
        shortSummary: 'Bla'
      }
    ]
  })

  const localVue = createLocalVue()
  localVue.use(VueRouter)

  const router = new VueRouter({
    routes: [
      { path: '/current/page', component: { render: () => '-' } }
    ]
  })

  const listWrapper = mount(List, {
    localVue,
    router,
    propsData: {
      apiEndpoint: '/books',
      cols: [
        new Column('title', 'Titre'),
        new Column('year', 'Année'),
        new Column('language', 'Langue'),
        new Column('authors', 'Auteurs')
          .setSearchParameterName('authorFullname')
          .setSubProperties([
            new DataSubProperty('firstname'),
            new DataSubProperty('lastname')
          ])
      ],
      leftActionBarProperties: new LeftActionBarProperties([
        new LeftActionBarElement(
          'separator',
          () => null,
          new LeftActionBarSeparatorDescriptor('add').setLabel('Ajouter').setFaIcon('fas fa-plus')
        ),
        new LeftActionBarElement(
          'element',
          () => null,
          new ButtonDescriptor('addPaper', 'Livre papier').setFaIcon('fas fa-scroll').setNoDefaultStyle(true)
        ),
        new LeftActionBarElement(
          'element',
          () => null,
          new ButtonDescriptor('addElectronic', 'Epub').setFaIcon('fas fa-tablet-alt').setNoDefaultStyle(true)
        ),
        new LeftActionBarElement(
          'separator',
          () => null,
          new LeftActionBarSeparatorDescriptor('filters').setLabel('Filtres').setFaIcon('fas fa-sliders-h')
        ),
        new LeftActionBarElement(
          'filter',
          () => null,
          new LeftActionBarFormSelectDescriptor('bookType', {
            all: 'Tous',
            paper: 'Papier',
            electronic: 'Epub'
          }).setDefaultValue('all').setFaIcon('fas fa-book').setNoDefaultStyle(true)
        )
      ], false),
      rowActions: [
        new RowAction('download', '', 'fas fa-file-download')
          .setIsDisplayed((book) => {
            return typeof book['@type'] === 'string' && book['@type'] === BookService.bookElectronic
          }),
        new RowAction('delete', '', 'far fa-trash-alt')
          .setNeedConfirm(true, 'Re-cliquez pour confirmer la suppression')
      ],
      detailsComponentPath: 'book/BookListRowDetails.vue'
    },
    data: () => {
      return {
        requestService: container.resolve(RequestService)
      }
    }
  })

  test('is a Vue instance', () => {
    expect(listWrapper.vm).toBeTruthy()
  })

  test('is rendered correctly', () => {
    expect(listWrapper.element).toMatchSnapshot()
  })
})
