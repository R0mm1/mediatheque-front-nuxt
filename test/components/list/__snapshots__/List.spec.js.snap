// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List is rendered correctly 1`] = `
<span
  id="vueListContainer"
>
  <div
    id="leftActionBar"
  >
    <div
      id="labColorBar"
    />
     
    <div
      id="labElements"
    >
      <div
        class="leftActionBarElement labe-leftActionBarSeparator"
      >
        <div
          class="leftActionBarElementIcon fas fa-plus"
        />
         
        <div
          class="leftActionBarElementText"
        >
          <!---->
           
          <!---->
           
          <div
            class="leftActionBarSeparatorText"
          >
            
          Ajouter
        
          </div>
        </div>
      </div>
      <div
        class="leftActionBarElement labe-button"
      >
        <div
          class="leftActionBarElementIcon fas fa-scroll"
        />
         
        <div
          class="leftActionBarElementText"
        >
          <div
            class="formulate-input"
            data-classification="button"
            data-type="button"
          >
            <div
              class="formulate-input-wrapper med_input_button style_normal no-def-style"
            >
              <!---->
               
              <!---->
               
              <div
                class="formulate-input-element formulate-input-element--button"
                data-type="button"
              >
                <button
                  id="formulate-global-1"
                  name="addPaper"
                  title="Livre papier"
                  type="button"
                >
                  <i
                    aria-hidden="false"
                    class="fas fa-scroll"
                  />
                  
  Livre papier

                </button>
              </div>
               
              <!---->
            </div>
             
            <!---->
             
            <!---->
          </div>
           
          <!---->
           
          <!---->
        </div>
      </div>
      <div
        class="leftActionBarElement labe-button"
      >
        <div
          class="leftActionBarElementIcon fas fa-tablet-alt"
        />
         
        <div
          class="leftActionBarElementText"
        >
          <div
            class="formulate-input"
            data-classification="button"
            data-type="button"
          >
            <div
              class="formulate-input-wrapper med_input_button style_normal no-def-style"
            >
              <!---->
               
              <!---->
               
              <div
                class="formulate-input-element formulate-input-element--button"
                data-type="button"
              >
                <button
                  id="formulate-global-2"
                  name="addElectronic"
                  title="Epub"
                  type="button"
                >
                  <i
                    aria-hidden="false"
                    class="fas fa-tablet-alt"
                  />
                  
  Epub

                </button>
              </div>
               
              <!---->
            </div>
             
            <!---->
             
            <!---->
          </div>
           
          <!---->
           
          <!---->
        </div>
      </div>
      <div
        class="leftActionBarElement labe-leftActionBarSeparator"
      >
        <div
          class="leftActionBarElementIcon fas fa-sliders-h"
        />
         
        <div
          class="leftActionBarElementText"
        >
          <!---->
           
          <!---->
           
          <div
            class="leftActionBarSeparatorText"
          >
            
          Filtres
        
          </div>
        </div>
      </div>
      <div
        class="leftActionBarElement labe-leftActionBarFormSelect"
      >
        <div
          class="leftActionBarElementIcon fas fa-book"
        />
         
        <div
          class="leftActionBarElementText"
        >
          <!---->
           
          <div
            class="formulate-input"
            data-classification="select"
            data-has-value="true"
            data-type="select"
          >
            <div
              class="formulate-input-wrapper med_select no-def-style"
            >
              <!---->
               
              <!---->
               
              <div
                class="formulate-input-element formulate-input-element--select"
                data-type="select"
              >
                <select
                  id="formulate-global-3"
                >
                  <!---->
                   
                  <option
                    value="all"
                  >
                    Tous
                  </option>
                  <option
                    value="paper"
                  >
                    Papier
                  </option>
                  <option
                    value="electronic"
                  >
                    Epub
                  </option>
                </select>
              </div>
               
              <!---->
            </div>
             
            <!---->
             
            <!---->
          </div>
           
          <!---->
        </div>
      </div>
    </div>
  </div>
   
  <div
    id="vueListContent"
  >
    <div
      class=""
      id="vueList"
    >
      <div
        class="list-header"
      >
        <div
          class="cell"
        >
          <div
            class="headerRow"
          >
            <div
              class="headerRowLabel"
            >
              
        Titre
      
            </div>
             
            <button
              class="headerTogglePopupButton fas fa-ellipsis-h"
            />
             
            <div
              class="buttonGroup"
            >
              <button
                class="headerSortButtonUp fas fa-sort-up"
              />
               
              <button
                class="headerSortButtonDown fas fa-sort-down"
              />
            </div>
          </div>
           
          <!---->
        </div>
        <div
          class="cell"
        >
          <div
            class="headerRow"
          >
            <div
              class="headerRowLabel"
            >
              
        Année
      
            </div>
             
            <button
              class="headerTogglePopupButton fas fa-ellipsis-h"
            />
             
            <div
              class="buttonGroup"
            >
              <button
                class="headerSortButtonUp fas fa-sort-up"
              />
               
              <button
                class="headerSortButtonDown fas fa-sort-down"
              />
            </div>
          </div>
           
          <!---->
        </div>
        <div
          class="cell"
        >
          <div
            class="headerRow"
          >
            <div
              class="headerRowLabel"
            >
              
        Langue
      
            </div>
             
            <button
              class="headerTogglePopupButton fas fa-ellipsis-h"
            />
             
            <div
              class="buttonGroup"
            >
              <button
                class="headerSortButtonUp fas fa-sort-up"
              />
               
              <button
                class="headerSortButtonDown fas fa-sort-down"
              />
            </div>
          </div>
           
          <!---->
        </div>
        <div
          class="cell"
        >
          <div
            class="headerRow"
          >
            <div
              class="headerRowLabel"
            >
              
        Auteurs
      
            </div>
             
            <button
              class="headerTogglePopupButton fas fa-ellipsis-h"
            />
             
            <div
              class="buttonGroup"
            >
              <button
                class="headerSortButtonUp fas fa-sort-up"
              />
               
              <button
                class="headerSortButtonDown fas fa-sort-down"
              />
            </div>
          </div>
           
          <!---->
        </div>
         
        <div
          class="cell"
        />
         
        <div
          class="list-header-scrollbar-width"
        />
      </div>
       
      <div
        id="vueListRows"
      >
        <div
          class="listRow"
          id="el39"
        >
          <div
            class="listCells"
          >
            <div
              class="cell"
            >
              
      Sharko
    
            </div>
            <div
              class="cell"
            >
              
      2017
    
            </div>
            <div
              class="cell"
            >
              
      Français
    
            </div>
            <div
              class="cell"
            >
              
      Franck Thilliez
    
            </div>
             
            <div
              class="listRowCustomActions cell"
            >
              <div
                class="rowCustomAction"
              >
                <div
                  class="formulate-input"
                  data-classification="button"
                  data-type="button"
                >
                  <div
                    class="formulate-input-wrapper med_input_button style_normal"
                  >
                    <!---->
                     
                    <!---->
                     
                    <div
                      class="formulate-input-element formulate-input-element--button"
                      data-type="button"
                    >
                      <button
                        id="formulate-global-4"
                        name="download"
                        title=""
                        type="button"
                      >
                        <i
                          aria-hidden="false"
                          class="fas fa-file-download"
                        />
                        
  

                      </button>
                    </div>
                     
                    <!---->
                  </div>
                   
                  <!---->
                   
                  <!---->
                </div>
                 
                <!---->
              </div>
              <div
                class="rowCustomAction"
              >
                <div
                  class="formulate-input"
                  data-classification="button"
                  data-type="button"
                >
                  <div
                    class="formulate-input-wrapper med_input_button style_normal"
                  >
                    <!---->
                     
                    <!---->
                     
                    <div
                      class="formulate-input-element formulate-input-element--button"
                      data-type="button"
                    >
                      <button
                        id="formulate-global-5"
                        name="delete"
                        title=""
                        type="button"
                      >
                        <i
                          aria-hidden="false"
                          class="far fa-trash-alt"
                        />
                        
  

                      </button>
                    </div>
                     
                    <!---->
                  </div>
                   
                  <!---->
                   
                  <!---->
                </div>
                 
                <div
                  class="customActionConfirm"
                >
                  
    Re-cliquez pour confirmer la suppression
  
                </div>
              </div>
            </div>
          </div>
           
          <!---->
        </div>
        <div
          class="listRow"
          id="el45"
        >
          <div
            class="listCells"
          >
            <div
              class="cell"
            >
              
      Sous les vents de Neptune
    
            </div>
            <div
              class="cell"
            >
              
      2004
    
            </div>
            <div
              class="cell"
            >
              
      Français
    
            </div>
            <div
              class="cell"
            >
              
      Fred Vargas
    
            </div>
             
            <div
              class="listRowCustomActions cell"
            >
              <!---->
              <div
                class="rowCustomAction"
              >
                <div
                  class="formulate-input"
                  data-classification="button"
                  data-type="button"
                >
                  <div
                    class="formulate-input-wrapper med_input_button style_normal"
                  >
                    <!---->
                     
                    <!---->
                     
                    <div
                      class="formulate-input-element formulate-input-element--button"
                      data-type="button"
                    >
                      <button
                        id="formulate-global-6"
                        name="delete"
                        title=""
                        type="button"
                      >
                        <i
                          aria-hidden="false"
                          class="far fa-trash-alt"
                        />
                        
  

                      </button>
                    </div>
                     
                    <!---->
                  </div>
                   
                  <!---->
                   
                  <!---->
                </div>
                 
                <div
                  class="customActionConfirm"
                >
                  
    Re-cliquez pour confirmer la suppression
  
                </div>
              </div>
            </div>
          </div>
           
          <!---->
        </div>
      </div>
    </div>
     
    <!---->
  </div>
</span>
`;
